src/java.base/share/classes/jdk/internal/util/ClassFileDumper.java

```diff
@@ -0,0 +1,233 @@
+        this.key = key;
+        this.dumpDir = path;
+    }
+
+    public String key() {
+        return key;
+    }
+    public boolean isEnabled() {
+        return dumpDir != null;
+    }
+
+    public Path dumpPath() {
+        return dumpDir;
+    }
+
+    public int incrementAndGetCounter() {
+        return counter.incrementAndGet();
+    }
+
+    public Path pathname(String internalName) {
+        return dumpDir.resolve(encodeForFilename(internalName) + ".class");
+    }
+
+    @SuppressWarnings("removal")
+    public void dumpClass(String internalName, final byte[] classBytes) {
+        if (!isEnabled()) return;
+
+        AccessController.doPrivileged(new PrivilegedAction<>() {
+                @Override public Void run() {
+                    Path file = pathname(internalName);
+                    try {
+                        Path dir = file.getParent();
+                        Files.createDirectories(dir);
+                        Files.write(file, classBytes);
+                    } catch (Exception ex) {
+                        if (VM.isModuleSystemInited()) {
+                            // log only when lambda is ready to use
+                            System.getLogger(ClassFileDumper.class.getName())
+                                  .log(System.Logger.Level.WARNING, "Exception writing to " +
+                                          file.toString() + " " + ex.getMessage());
+                        }
+                        // simply don't care if this operation failed
+                    }
+                    return null;
+                }},
+                null,
+                new FilePermission("<<ALL FILES>>", "read, write"),
+                // createDirectories may need it
+                new PropertyPermission("user.dir", "read"));
+    }
+
+    @SuppressWarnings("removal")
+    private static Path validateDumpDir(Path path) {
+            return AccessController.doPrivileged(new PrivilegedAction<>() {
+                @Override
+                public Path run() {
+                    if (!Files.exists(path)) {
+                        try {
+                            Files.createDirectory(path);
+                        } catch (IOException ex) {
+                            throw new UncheckedIOException("Fail to create " + path, ex);
+                        }
+                    }
+                    if (!Files.isDirectory(path)) {
+                        throw new IllegalArgumentException("Path " + path + " is not a directory");
+                    } else if (!Files.isWritable(path)) {
+                        throw new IllegalArgumentException("Directory " + path + " is not writable");
+                    }
+                    return path;
+                }
+            });
+    }
+
+
+    private static String encodeForFilename(String className) {
+        final int len = className.length();
+        StringBuilder sb = new StringBuilder(len);
+
```

Comment:
This could be fewer lines of adhoc conversions, the replacements are hex encoding of the characters, so the replacement array is not needed. 
java.util.HexFormat should be used for the hex encoding. The `formatHex` methods append to a StringBuilder.

Reply:
This old code was from ProxyClassesDumper.  Yes it can be converted to use HexFormat.

Reply:
> This could be fewer lines of adhoc conversions, the replacements are hex encoding of the characters, so the replacement array is not needed. 
> 
> java.util.HexFormat should be used for the hex encoding. The `formatHex` methods append to a StringBuilder.